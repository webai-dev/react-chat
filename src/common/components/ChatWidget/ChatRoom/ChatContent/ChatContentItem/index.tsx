import React, { useContext, useState, useEffect } from "react";
import { makeStyles } from "@material-ui/styles";
import { Box, Grid, Container } from "@material-ui/core";
import { MeContext } from "../../../../../hooks/useMe";
import { ChatMessage, Emojis } from "../../../../../../types/types";
import { toArray } from "react-emoji-render";
import EmojiEmotionsIcon from "@material-ui/icons/EmojiEmotions";
import ReplyIcon from "@material-ui/icons/Reply";
import "emoji-mart/css/emoji-mart.css";
import { Picker, EmojiData, Emoji } from "emoji-mart";
import OutsideClickHandler from "react-outside-click-handler";
import CheckIcon from "@material-ui/icons/Check";
import Tooltip from "@material-ui/core/Tooltip";
import Linkify from "react-linkify";
import ErrorIcon from "@material-ui/icons/ErrorOutline";
import { v4 as uuidv4 } from "uuid";

import ChatService from "../../../../../services/chat";
import { CompanyContext } from "../../../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../../../hooks/useChatRooms";
import {NewMessage} from "../../../../../../types";
// import ChatRoom from "../..";

// import { Emoji, Emoji } from "emoji-mart/dist-es/utils/data";

// Global styles
export const useStyles = makeStyles({
  root: {
    display: "flex",
    flexDirection: "column",
    alignSelf: (sent) => (sent ? "flex-end" : "flex-start"),
    width: "220px",
    padding: "5px",
    paddingBottom: 0,
    marginLeft: (sent) => (sent ? "auto" : 0),
    boxShadow: "0px 1px 3px #00000029",
    background: (sent) =>
      sent
        ? "#E0E0E8 0% 0% no-repeat padding-box"
        : "#F3F6F8 0% 0% no-repeat padding-box",
  },
  picker: {
    position: "absolute",
    right: "80px",
    bottom: "20px",
    zIndex: 3,
  },
  emojisContainer: {
    display: "flex",
    width: "30px",
    justifyContent: "space-between",
    alignItems: "center",
    padding: "3px",
    backgroundColor: "rgba(50, 50, 50, .1)",
    borderRadius: "10px",
    cursor: "pointer",
  },
  emojisContainerMe: {
    display: "flex",
    width: "30px",
    justifyContent: "space-between",
    alignItems: "center",
    padding: "3px",
    backgroundColor: "rgba(29, 155, 209, .2)",
    border: "1px solid rgba(29, 155, 209, .2)",
    borderRadius: "10px",
    cursor: (sent) => (sent ? "auto" : "pointer"),
  },
  emojis: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "flex-end",
  },
  emojisCount: {
    margin: 0,
    fontSize: "12px",
    color: "rgba(50, 50, 50, 1)",
  },
  emojisCountMe: {
    margin: 0,
    fontSize: "12px",
    color: "rgba(29, 155, 209, 1)",
  },
  failedMessage: {
    fontSize: 12,
    color: "#f00",
    textAlign: "right",
    marginTop: 5,
  },
  icon: {
    fontSize: "16px",
    color: "rgb(164, 175, 183)",
    cursor: "pointer",
    padding: "5px 5px",
  },
  retryIcon: {
    fontSize: 36,
    color: "rgb(255, 0, 0)",
    cursor: "pointer",
    paddingTop: "15px",
    height: "100%",
    marginTop: "auto",
    marginBottom: "auto",
  },
  iconContainer: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    "&:hover": {
      backgroundColor: "#eceff4",
    },
  },
  userContainer: {
    display: "flex",
    justifyContent: "space-between",
    paddingTop: "5px",
    height: "21px",
    paddingBottom: "3px",
  },
  messageContainer: {
    display: "flex",
    justifyContent: "space-between",
    alignItems: "center",
    paddingTop: "2px",
    paddingBottom: "3px",
  },
  user: {
    fontSize: "10px",
    color: "#555555",
    fontWeight: "bold",
    letterSpacing: "0px",
    textAlign: "left",
    paddingRight: "15px",
  },
  time: {
    fontSize: "10px",
    color: "#555555",
    letterSpacing: "0px",
    textAlign: "left",
  },
  message: {
    fontSize: "12px",
    color: "#555555",
    letterSpacing: "0px",
    paddingBottom: "5px",
    textAlign: "left",
  },
  messageErrorPopup: {
    position: "absolute",
    backgroundColor: "white",
    boxShadow: "rgba(100, 100, 111, 0.2) 0px 7px 29px 0px",
    width: "80px",
    height: "50px",
    fontSize: "14px",
    right: 0,
  },
  reactionsContainer: {
    display: "flex",
    flexDirection: "row",
    alignItems: "flex-start",
    justifyContent: (sent) => (sent ? "flex-end" : "flex-start"),
  },
  reaction: {
    padding: "5px 3px",
  },
  checkGray: {
    fontSize: "14px !important",
    color: "#757373",
    letterSpacing: "0px",
    textAlign: "right",

    display: "flex",
    flexDirection: "column",
    alignItems: "flex-end",
    justifyContent: "flex-end",
  },
  check: {
    fontSize: "14px !important",
    color: "#3eaf3e",
    letterSpacing: "0px",
    textAlign: "right",

    display: "flex",
    flexDirection: "column",
    alignItems: "flex-end",
    justifyContent: "flex-end",
  },
});

interface ChatContentItemProps {
  message: ChatMessage | null;
  roomId?: number;
  companyId?: number | null;
  connectionStatus: number;
  singleReadPosition: number | undefined;
  groupAllReadPosition: number;
  groupReadPosition: any;
  roomType: number | undefined;
}

/**
 * ChatContentItem component
 * @param {ChatContentItemProps} props Properties
 * @description ChatContentItem component consists of messages UI
 */
const ChatContentItem: React.FC<ChatContentItemProps> = ({
  message = null,
  roomId = 0,
  companyId = null,
  connectionStatus = 1,
  singleReadPosition = -1,
  groupReadPosition = {},
  groupAllReadPosition = -1,
  roomType = 0,
}) => {
  // Default styling
  const { me } = useContext(MeContext);

  const classes = useStyles(me?.id === message?.user?.id);
  const { addReaction, removeReaction, reactionUser, sendMessage } =
    ChatService;
  const [emojisList, setEmojisList] = useState<Array<Emojis>>([]);
  const [emojiUsers, setEmojiUsers] = useState<Array<string>>([]);
  const [newReactions, setNewReactions] = useState<Array<string>>([]);
  const [showEmojis, setShowEmojis] = useState<boolean>(false);
  const [showPicker, setShowPicker] = useState<boolean>(false);
  const [failed, setFailed] = useState<boolean>(false);
  const [messageCheck, setMessageCheck] = useState<string | undefined>("");
  const [isSentOnce, setIsSentOnce] = useState<boolean>(false);
  const { selectedCompany } = useContext(CompanyContext);
  const { removeFailedMessage } = useContext(ChatRoomsContext);

  const oneDayAgo = (date: string) => {
    const day = 1000 * 60 * 60 * 12;
    const daysago = Date.now() - day;

    return new Date(date).getTime() > daysago;
  };

  const getSentDate = (date: string) => {
    if (oneDayAgo(date)) {
      return new Date(date).toLocaleTimeString([], {
        hour: "2-digit",
        minute: "2-digit",
      });
    } else {
      return new Date(date).toLocaleString([], {
        year: "numeric",
        month: "numeric",
        day: "numeric",
        hour: "2-digit",
        minute: "2-digit",
      });
    }
  };

  const getReactionUser = async (emoji: string) => {
    setEmojiUsers([]);
    let users = await reactionUser(companyId, roomId, message?.id, emoji);
    let userName: Array<string> = [];
    users.forEach((u) => userName.push(u.firstName ? u.firstName : ""));
    setEmojiUsers(userName);
  };

  useEffect(() => {
    setMessageCheck(message?.check);
    !isSentOnce && setIsSentOnce(message?.check === "sent");
    let reactions: Array<Emojis> = [];
    if (message) {
      message?.reactions &&
        Object.entries(message?.reactions).forEach(([key, value]) => {
          let userName: Array<string> = [];

          reactions.push({
            name: key,
            count: value ? parseInt(value) : 1,
            userName: userName,
          });
        });
      setEmojisList(reactions);
    }
  }, [message, message?.reactions]);

  const onReactionClick = () => {
    setShowPicker(!showPicker);
  };

  const handleEmojiSelect = async (emoji: EmojiData) => {
    if (emojisList.some((e) => e.name === emoji.id)) {
      if (
        message?.reactions &&
        Object.entries(message?.reactions).some(
          ([key, value]) => key === emoji.id
        ) &&
        emoji.id
      ) {
        if (
          (me &&
            message?.userReactions &&
            !message.userReactions.includes(emoji.id)) ||
          (me && !message.userReactions && !newReactions.includes(emoji.id))
        ) {
          let newEmojisList = emojisList;
          let index = emojisList.findIndex((e) => e.name === emoji.id);

          let newCount = emojisList[index].count;
          newEmojisList[index] = {
            ...emojisList[index],
            count: newCount ? newCount + 1 : 1,
          };
          setEmojisList(newEmojisList);

          let reaction = { reaction: emoji.id };
          if (
            emoji.id &&
            message?.userReactions &&
            !message?.userReactions?.includes(emoji.id)
          ) {
            message?.userReactions?.push(emoji.id);
          } else if (message && emoji.id) {
            setNewReactions([...newReactions, emoji.id]);
          }
          setShowPicker(false);

          await addReaction(companyId, roomId, message?.id, reaction);
        } else {
          let newEmojisList = emojisList;
          let index = emojisList.findIndex((e) => e.name === emoji.id);
          let newCount = emojisList[index].count;
          if (newCount && newCount > 1) {
            newEmojisList[index] = {
              ...emojisList[index],
              count: newCount - 1,
            };
            setEmojisList(newEmojisList);
          } else {
            setEmojisList(emojisList.filter((e) => e.name !== emoji.id));
          }

          if (emoji.id && message.userReactions) {
            message.userReactions = message.userReactions?.filter(
              (m) => m !== emoji.id
            );
          }
          setNewReactions(newReactions.filter((r) => r !== emoji.id));

          let reaction = { reaction: emoji.id };
          await removeReaction(companyId, roomId, message?.id, reaction);
        }
      } else {
        let newEmojisList = emojisList;
        let index = emojisList.findIndex((e) => e.name === emoji.id);
        let newCount = emojisList[index].count;
        if (newCount && newCount > 1) {
          newEmojisList[index] = {
            ...emojisList[index],
            count: newCount - 1,
          };
          setEmojisList(newEmojisList);
        } else {
          setEmojisList(emojisList.filter((e) => e.name !== emoji.id));
        }
      }

      setShowPicker(false);
    } else {
      setEmojisList([
        ...emojisList,
        {
          count: 1,
          name: emoji.id ? emoji.id : "",
          userName: [me ? me?.firstName : ""],
        },
      ]);
      let reaction = { reaction: emoji.id };
      if (
        emoji.id &&
        message?.userReactions &&
        !message?.userReactions?.includes(emoji.id)
      ) {
        message?.userReactions?.push(emoji.id);
      } else if (message && emoji.id) {
        setNewReactions([...newReactions, emoji.id]);
      }

      setShowPicker(false);

      await addReaction(companyId, roomId, message?.id, reaction);
    }
  };

  const handleEmojiClick = async (emoji: Emojis) => {
    if (emojisList.some((e) => e.name === emoji.name)) {
      if (
        message?.reactions &&
        Object.entries(message?.reactions).some(
          ([key, value]) => key === emoji.name
        ) &&
        emoji.name
      ) {
        if (
          (me &&
            message?.userReactions &&
            !message.userReactions.includes(emoji.name)) ||
          (me && !message.userReactions && !newReactions.includes(emoji.name))
        ) {
          let newEmojisList = emojisList;
          let index = emojisList.findIndex((e) => e.name === emoji.name);
          let newCount = emojisList[index].count;
          newEmojisList[index] = {
            ...emojisList[index],
            count: newCount ? newCount + 1 : 1,
          };
          setEmojisList(newEmojisList);

          let reaction = { reaction: emoji.name };
          if (
            emoji.name &&
            message?.userReactions &&
            !message?.userReactions?.includes(emoji.name)
          ) {
            message?.userReactions?.push(emoji.name);
          } else if (message && emoji.name) {
            setNewReactions([...newReactions, emoji.name]);
          }
          setShowPicker(false);

          await addReaction(companyId, roomId, message?.id, reaction);
        } else {
          let newEmojisList = emojisList;
          let index = emojisList.findIndex((e) => e.name === emoji.name);
          let newCount = emojisList[index].count;
          if (newCount && newCount > 1) {
            newEmojisList[index] = {
              ...emojisList[index],
              count: newCount - 1,
            };
            setEmojisList(newEmojisList);
          } else {
            setEmojisList(emojisList.filter((e) => e.name !== emoji.name));
          }

          if (emoji.name && message.userReactions) {
            message.userReactions = message.userReactions?.filter(
              (m) => m !== emoji.name
            );
          }
          setNewReactions(newReactions.filter((r) => r !== emoji.name));

          let reaction = { reaction: emoji.name };
          await removeReaction(companyId, roomId, message?.id, reaction);
        }
      } else {
        let newEmojisList = emojisList;
        let index = emojisList.findIndex((e) => e.name === emoji.name);
        let newCount = emojisList[index].count;
        if (newCount && newCount > 1) {
          newEmojisList[index] = {
            ...emojisList[index],
            count: newCount - 1,
          };
          setEmojisList(newEmojisList);
        } else {
          setEmojisList(emojisList.filter((e) => e.name !== emoji.name));
        }
        let reaction = { reaction: emoji.name };
        await removeReaction(companyId, roomId, message?.id, reaction);
      }

      setShowPicker(false);
    }
  };

  const parseEmojis = (value: string | undefined) => {
    if (value) {
      const emojisArray = toArray(value);

      // toArray outputs React elements for emojis and strings for other
      const newValue = emojisArray.reduce((previous, current: any) => {
        if (typeof current === "string") {
          return previous + current;
        }
        return previous + current?.props?.children;
      }, "");

      return newValue;
    }
    return "";
  };

  const handleSend = async () => {
    let data = {
      message: message?.message,
      trackingID: uuidv4(),
      sentAt: new Date(),
    };

    if(message) {
      setMessageCheck("sent");
      setFailed(false);
    }

    if (connectionStatus === 1) {
      const response = await sendMessage(
          selectedCompany && selectedCompany.id,
          roomId,
          data
      );
      if (!response) {
        setMessageCheck("failed");
      }
    } else {
      let response: NewMessage | null = null;
      if (!(connectionStatus === 1)) {
        response = await sendMessage(
            selectedCompany && selectedCompany.id,
            roomId,
            data
        );

        if (response) {
          setMessageCheck("received");
          setIsSentOnce(false);
        } else {
          let started = Date.now();
          let interval = setInterval(async function () {
            if (Date.now() - started > 46000) {
              clearInterval(interval);
              setMessageCheck("failed");
              setFailed(false);
            }
            // Retry sending message for first 2 minutes or after 5 minutes
            else {
              // let response = connectionStatus === 1 ? 1 : 0;
              if (response) {
                setMessageCheck("received");
                setIsSentOnce(false);
                clearInterval(interval);
              } else {
                 response = await sendMessage(
                    selectedCompany && selectedCompany.id,
                    roomId,
                    data
                );
              }
            }
          }, 15000);
        }
      }
    }
  };

  return (
    <div
      onMouseEnter={() => {
        setShowEmojis(true);
      }}
      onMouseLeave={() => {
        setShowEmojis(false);
      }}
    >
      <div
        style={{
          display: "flex",
          flexDirection: "row",
        }}
      >
        <Grid
          data-testid="chat-content-item"
          container
          justify="flex-end"
          className={classes.root}
        >
          <div className={classes.userContainer}>
            <div className={classes.user}>{message?.user?.firstName}</div>
            <div className={classes.time}>
              {getSentDate(message?.sentAt || "")}
            </div>
            {showEmojis && me?.id !== message?.user?.id && (
              <span
                style={{
                  display: "flex",
                  flexDirection: "row",
                  justifyContent: "center",
                  alignItems: "center",
                }}
              >
                <span>
                  {" "}
                  <EmojiEmotionsIcon
                    style={{
                      fontSize: "16px",
                      color: "rgb(164, 175, 183)",
                      cursor: "pointer",
                      paddingLeft: "10px",
                    }}
                    onClick={onReactionClick}
                  />
                </span>

                <span>
                  {" "}
                  <ReplyIcon
                    style={{
                      fontSize: "16px",
                      color: "rgb(164, 175, 183)",
                      cursor: "pointer",
                      paddingLeft: "10px",
                    }}
                  />
                </span>
              </span>
            )}
          </div>
          <div className={classes.messageContainer}>
            <Linkify>
              <div className={classes.message}>
                {parseEmojis(message?.message)}
              </div>
            </Linkify>
            {
              message?.user.id === me?.id && (
                <div>
                  {(message?.check === "sent" || isSentOnce && message?.check !== "received" && messageCheck !== "received") && (
                    <CheckIcon className={classes.checkGray} />
                  )}
                  {
                    message && (
                      <React.Fragment>
                        {
                          ((roomType === 0 && message.id > groupAllReadPosition) || message?.check === "received") &&
                          <CheckIcon className={classes.check} />
                        }
                        {
                          roomType !== 0 && message.id > singleReadPosition && <CheckIcon className={classes.check} />
                        }
                      </React.Fragment>
                    )
                  }
                </div>
              )
            }
          </div>
        </Grid>
        {me?.id === message?.user?.id && messageCheck === "failed" && (
          <Box position="relative" height="100%">
            <ErrorIcon
              onClick={() => {
                setFailed(!failed);
              }}
              className={classes.retryIcon}
            />

            {failed && (
              <Container className={classes.messageErrorPopup}>
                {" "}
                <div
                  style={{
                    paddingTop: "5px",
                    marginBottom: "5px",
                    cursor: "pointer",
                  }}
                  onClick={() => {
                    handleSend();
                  }}
                >
                  Retry
                </div>{" "}
                <div
                  style={{ paddingTop: "5px", cursor: "pointer" }}
                  onClick={() => {
                    removeFailedMessage(roomId, message?.trackingID || "");
                  }}
                >
                  Delete
                </div>{" "}
              </Container>
            )}
          </Box>
        )}
      </div>
      {me?.id === message?.user?.id && messageCheck === "failed" && (
        <div className={classes.failedMessage}>Failed to Send</div>
      )}
      <div className={classes.reactionsContainer}>
        {emojisList.map((emoji, index) => {
          return (
            <Tooltip
              key={index}
              disableFocusListener
              title={
                emojiUsers.length > 1
                  ? `${emojiUsers.slice(0, -1).join(", ")} and ${
                      emojiUsers[emojiUsers.length - 1]
                    } reacted with :${emoji.name}:`
                  : emojiUsers.length === 1
                  ? `${emojiUsers[0]} reacted with :${emoji.name}:`
                  : ""
              }
              placement="top-start"
              onMouseEnter={() => {
                getReactionUser(emoji.name);
              }}
              arrow
            >
              <div
                className={classes.reaction}
                onClick={() => {
                  if (me && message?.user.id !== me.id) handleEmojiClick(emoji);
                }}
              >
                <div id={`${index}`}>
                  <div
                    className={
                      (message?.userReactions &&
                        message.userReactions.includes(emoji.name)) ||
                      newReactions.includes(emoji.name)
                        ? classes.emojisContainerMe
                        : classes.emojisContainer
                    }
                  >
                    {" "}
                    <Emoji emoji={emoji.name} size={16} />{" "}
                    <p
                      className={
                        (message?.userReactions &&
                          message.userReactions.includes(emoji.name)) ||
                        newReactions.includes(emoji.name)
                          ? classes.emojisCountMe
                          : classes.emojisCount
                      }
                    >
                      {emoji.count}
                    </p>
                  </div>
                </div>
              </div>
            </Tooltip>
          );
        })}
      </div>

      {showPicker && (
        <OutsideClickHandler
          onOutsideClick={() => {
            setShowPicker(false);
          }}
        >
          {" "}
          <span className={classes.picker}>
            <Picker
              perLine={5}
              showPreview={false}
              showSkinTones={false}
              onSelect={handleEmojiSelect}
              set="apple"
            />
          </span>
        </OutsideClickHandler>
      )}
    </div>
  );
};

export default ChatContentItem;
