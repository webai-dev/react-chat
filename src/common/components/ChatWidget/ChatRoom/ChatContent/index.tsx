import React, { useRef, useState, useEffect, useContext } from "react";
import { makeStyles } from "@material-ui/styles";
import ChatContentItem from "./ChatContentItem";
import Box from "@material-ui/core/Box";
import ChatService from "../../../../services/chat";
import { ActiveUser, ChatRoom, RoomDetail } from "../../../../../types/types";
import { CompanyContext } from "../../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../../hooks/useChatRooms";
import { RoomListsContext } from "../../../../hooks/useRoomLists";
import { getRoomDetail } from "../../../../services/rooms";
import { MeContext } from "../../../../hooks/useMe";

import One2OneReadSign from "./One2OneReadSign";
import GroupReadSign from "./GroupReadSign";

// Global styles
export const useStyles = makeStyles({
  root: {
    overflowY: "auto",
    justifyContent: "flex-end",
    height: "320px",
    padding: "5px 15px",
    bottom: "75px",
  },
  loading: {
    fontSize: "12px",
    color: "#555555",
    letterSpacing: "0px",
    textAlign: "center",
  },
  loadingWrapper: {
    position: "absolute",
    top: "0",
    left: "0",
    width: "100%",
    height: "100%",
    backgroundColor: "white",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  reactionsLeft: {
    display: "flex",
    flexDirection: "column",
    alignSelf: "flex-start",
  },
  reactionsRight: {
    display: "flex",
    flexDirection: "column",
    alignSelf: "flex-end",
  },
  chatItemWrapper: {
    paddingTop: 15,
  },
  readUsers: {
    marginTop: "5px",
    marginRight: "5px",
    width: "100%",
    height: "14px",
    display: "flex",
    justifyContent: "flex-end",
    fontSize: 13,
    cursor: 'pointer',
  },
  readUsersImage: {
    borderRadius: "50%",
    border: "1px solid #cccccc",
  },
});

interface ChatContentProps {
  chatRoom: ChatRoom | null | undefined;
  connectionStatus: number;
}

/**
 * ChatContent component
 * @param {ChatContentProps} props Properties
 * @description ChatContent component consists of list of ChatContentItem components
 */
const ChatContent: React.FC<ChatContentProps> = ({
  chatRoom = null,
  connectionStatus = 1,
}) => {
  // Default styling
  const classes = useStyles();
  const { markAsRead } = ChatService;
  const { companies, setCompanies } =
    useContext(CompanyContext);
  const { loadPrevPage, loadNextPage, initializeChatRoom, updateChatRooms } =
    useContext(ChatRoomsContext);
  const { me } = useContext(MeContext);
  const {
    activeRoomList,
    companyRoomLists,
    getRoomItemById,
    updateCompanyRoomLists,
  } = useContext(RoomListsContext);
  const messageRoomRef = useRef() as React.MutableRefObject<HTMLInputElement>;

  const [detailLoading, setDetailLoading] = useState<boolean>(false);

  useEffect(() => {
    if (messageRoomRef && messageRoomRef.current && chatRoom && chatRoom.newMessageFromMe) {
      chatRoom.newMessageFromMe = false;
      messageRoomRef.current.scrollTop = messageRoomRef.current.scrollHeight;
      updateChatRooms(chatRoom);
    }
  }, [chatRoom?.newMessageFromMe, messageRoomRef]);

  useEffect(() => {
    if (messageRoomRef && messageRoomRef.current && chatRoom && chatRoom.newMessageFromOther) {
      chatRoom.newMessageFromOther = false;
      if (chatRoom.visible && chatRoom.focused && chatRoom.expanded && messageRoomRef.current.scrollTop + messageRoomRef.current.clientHeight + 73 > messageRoomRef.current.scrollHeight - 10) {
        messageRoomRef.current.scrollTop = messageRoomRef.current.scrollHeight;
      }
      updateChatRooms(chatRoom);
    }
  }, [chatRoom?.newMessageFromOther, messageRoomRef]);

  useEffect(() => {
    if (chatRoom && messageRoomRef.current) {
      if (messageRoomRef.current.scrollTop + messageRoomRef.current.clientHeight > messageRoomRef.current.scrollHeight - 10) {
        chatRoom.scrollAtBottom = true;
      } else {
        chatRoom.scrollAtBottom = false;
      }
      updateChatRooms(chatRoom);

    }
  }, [chatRoom, messageRoomRef.current?.scrollTop]);

  useEffect(() => {
    if (me && chatRoom && !chatRoom.initialized) {
      initializeChatRoom(chatRoom.roomId).then(() => {
        setDetailLoading(true);
        getRoomDetail(chatRoom.companyId, chatRoom.roomId).then(
          (data: RoomDetail | null) => {
            chatRoom.participants = data?.participants;
            chatRoom.roomType = data?.roomType;

            const participant = chatRoom.participants?.find(p => p.id === me?.id);
            const lastMessage = chatRoom.messages[chatRoom.messages.length - 1];

            if(participant && lastMessage && participant.lastReadMessageID === lastMessage.id) {
              messageRoomRef.current.scrollTop = messageRoomRef.current.scrollHeight;
              chatRoom.scrollPosition = messageRoomRef.current.scrollHeight;
            } else if (participant) {
              const message = document.getElementById(
                  `message-${participant.lastReadMessageID}`
              );

              if (message && messageRoomRef.current) {
                messageRoomRef.current.scrollTop = message.offsetTop;
                chatRoom.scrollPosition = message.offsetTop;
              }
            }

            updateChatRooms(chatRoom);

            if (messageRoomRef.current) {
              const { scrollHeight } = messageRoomRef.current;
              if (scrollHeight <= 330 && chatRoom) {
                markRoomRead(chatRoom);
              }
            }
            setDetailLoading(false);
          }
        );
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chatRoom, me]);

  useEffect(() => {
    if (chatRoom && chatRoom.initialized) {
      messageRoomRef.current.scrollTop = chatRoom.scrollPosition || 0;
    } // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chatRoom]);

  const markRoomRead = async (chatRoom: ChatRoom) => {
    const company = companies.find((c) => c.id === chatRoom.companyId);
    const room = companyRoomLists
      .map((roomList) => roomList.rooms)
      .flat()
      .find((room) => room && room.id === chatRoom.roomId);

    if (chatRoom.focused && chatRoom.expanded && chatRoom.visible) {
      if (activeRoomList && activeRoomList.rooms) {
        const room = activeRoomList.rooms.find(
          (room) => room && room.id === chatRoom.roomId
        );

        if (room && chatRoom.messages && chatRoom.messages.length > 0) {
          room.lastReadMessageID = chatRoom.messages[
            chatRoom.messages.length - 1
          ]
            ? chatRoom.messages[chatRoom.messages.length - 1].id
            : -1;
          room.unreadCount = 0;
        }

        updateCompanyRoomLists({ ...activeRoomList });
      }
      if (company && room) {
        let totalUnreadCount = 0;
        companyRoomLists.forEach((room) => {
          if (room.companyId === company.id) {
            room.rooms &&
              room.rooms.forEach((r) => {
                if (r.unreadCount > 0) {
                  totalUnreadCount += 1;
                }
              });
          }
        });

        company.unreadCount = totalUnreadCount;
        localStorage.setItem("companies", JSON.stringify([...companies]));
        setCompanies([...companies]);
      }
      await markAsRead(chatRoom.companyId, chatRoom.roomId);
    }
  };

  const handleScroll = async () => {
    if (chatRoom && messageRoomRef.current) {
      const { scrollTop, clientHeight, scrollHeight } = messageRoomRef.current;

      chatRoom.scrollPosition = scrollTop;
      if (chatRoom.prevHasMore && !chatRoom.loadingPrev && scrollTop === 0) {
        loadPrevPage(chatRoom.roomId).then((messages) => {
          if (messageRoomRef.current) {
            messageRoomRef.current.scrollTop = messages.length * 63 + 1;
          }
        });
      } else if (
        chatRoom.nextHasMore &&
        !chatRoom.loadingNext &&
        scrollTop + clientHeight === scrollHeight
      ) {
        const room = getRoomItemById(chatRoom.roomId);
        if (
          chatRoom.messages[chatRoom.messages.length - 1] &&
          room?.lastReadMessageID !==
            chatRoom.messages[chatRoom.messages.length - 1].id
        ) {
          const company = companies.find(
            (c) => c && c.id === chatRoom.companyId
          );
          const room = companyRoomLists
            .map((roomList) => roomList.rooms)
            .flat()
            .find((room) => room && room.id === chatRoom.roomId);
          if (chatRoom.focused && chatRoom.expanded && chatRoom.visible) {
            if (activeRoomList && activeRoomList.rooms) {
              const room = activeRoomList.rooms.find(
                (room) => room && room.id === chatRoom.roomId
              );

              if (room) {
                room.lastReadMessageID = chatRoom.messages[
                  chatRoom.messages.length - 1
                ]
                  ? chatRoom.messages[chatRoom.messages.length - 1].id
                  : -1;
                room.unreadCount = 0;
              }

              updateCompanyRoomLists({ ...activeRoomList });
            }
            if (company && room) {
              let totalUnreadCount = 0;
              companyRoomLists.forEach((room) => {
                if (room.companyId === company.id) {
                  room.rooms &&
                    room.rooms.forEach((r) => {
                      if (r && r.unreadCount > 0) {
                        totalUnreadCount += 1;
                      }
                    });
                }
              });

              company.unreadCount = totalUnreadCount;

              localStorage.setItem("companies", JSON.stringify([...companies]));
              setCompanies([...companies]);
            }

            await markAsRead(chatRoom.companyId, chatRoom.roomId);
          }
        }

        loadNextPage(chatRoom.roomId).then((messages) => {
          if (messageRoomRef.current) {
            messageRoomRef.current.scrollTop =
              messageRoomRef.current.scrollHeight - messages.length * 63 - 1;
          }
        });
        chatRoom.messages.forEach((message) => (message.new = false));
      } else if (scrollTop + clientHeight === scrollHeight) {
        markRoomRead(chatRoom);
        // await markAsRead(chatRoom.companyId, chatRoom.roomId);
      }
    }
  };

  const participant = chatRoom?.participants?.find((participant: ActiveUser) => participant.id !== me?.id);
  const singleReadPosition = participant && chatRoom?.roomType !== 0 ? participant.lastReadMessageID : -1;
  let groupReadPosition: any = {};
  let groupAllReadPosition = -1;

  if (chatRoom && chatRoom.roomType === 0) {
    chatRoom.participants?.forEach((participant: ActiveUser) => {
      const messageID = participant.lastReadMessageID ? participant.lastReadMessageID.toString() : '-1';

      if (messageID !== '-1' && participant.id !== me?.id) {
        if (groupReadPosition[messageID]) {
          groupReadPosition[messageID].push(participant.id);
        } else {
          groupReadPosition[messageID] = [participant.id];
        }
      }
    });

    const sortedIds = Object.keys(groupReadPosition).map(id => parseInt(id)).sort((a, b) => a > b ? 1 : -1);
    let sum = 0;

    for (let i = 0; i < sortedIds.length; i++) {
      sum += groupReadPosition[sortedIds[i]].length;
      if (sum === ((chatRoom.participants?.length || 1) - 2)) {
        groupAllReadPosition = sortedIds[i];
        break;
      }
    }
  }

  return (
    <>
      <div
        data-testid="chat-content"
        className={classes.root}
        ref={messageRoomRef}
        onScroll={handleScroll}
      >
        {(chatRoom?.loadingPrev && detailLoading) && (
          <Box
            display="flex"
            flex={1}
            height={90}
            justifyContent="center"
            alignItems="center"
          >
            <p className={classes.loading}>Loading history...</p>
          </Box>
        )}
        {chatRoom &&
          (chatRoom.messages || [])
            .map((message, index) => (
              <React.Fragment>
                <div
                  id={`message-${message.id || message.trackingID}`}
                  key={index}
                  className={classes.chatItemWrapper}
                >
                  {
                     detailLoading && (
                        <Box className={classes.loadingWrapper}>
                          <p className={classes.loading}>Loading history...</p>
                        </Box>
                     )
                  }
                  <ChatContentItem
                    message={message}
                    connectionStatus={connectionStatus}
                    roomId={chatRoom.roomId}
                    companyId={chatRoom.companyId ? chatRoom.companyId : null}
                    singleReadPosition={singleReadPosition}
                    groupReadPosition={groupReadPosition}
                    groupAllReadPosition={groupAllReadPosition}
                    roomType={chatRoom.roomType}
                  />
                </div>
                {
                  (chatRoom?.roomType !== 0 && message.id === singleReadPosition) && (
                    <One2OneReadSign
                      chatRoom={chatRoom}
                      me={me}
                    />
                  )
                }
                {
                  (chatRoom?.roomType === 0 && groupReadPosition[message.id]) && (
                    <GroupReadSign
                      participants={chatRoom?.participants}
                      groupReadPosition={groupReadPosition}
                      message={message}
                    />
                  )
                }
              </React.Fragment>
            ))}
        {chatRoom?.loadingNext && (
          <Box
            display="flex"
            flex={1}
            height={90}
            justifyContent="center"
            alignItems="center"
          >
            <p className={classes.loading}>Loading history...</p>
          </Box>
        )}
      </div>
    </>
  );
};

export default ChatContent;
