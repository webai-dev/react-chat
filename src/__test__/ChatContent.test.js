import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import ChatContent from "../common/components/ChatWidget/ChatRoom/ChatContent";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const { getByTestId } = render(<ChatContent />);
  const chatContent = getByTestId("chat-content");
  expect(chatContent).toBeTruthy();
});

it("matches snapshot", () => {
  const tree = renderer.create(<ChatContent />).toJSON();
  expect(tree).toMatchSnapshot();
});
