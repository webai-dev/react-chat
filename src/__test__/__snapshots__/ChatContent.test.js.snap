// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="makeStyles-root-8"
  data-testid="chat-content"
  onScroll={[Function]}
>
  <br />
</div>
`;
